grblas/vector.py

```diff
@@ -20,21 +20,70 @@ def __del__(self):
     def __repr__(self):
         return f'<Vector {self.nvals}/{self.size}:{self.dtype.name}>'
 
-    def __eq__(self, other):
-        # Borrowed this recipe from LAGraph
+    def isequal(self, other, *, check_dtype=False):
+        """
+        Check for exact equality (same size, same empty values)
+        If `check_dtype` is True, also checks that dtypes match
+        For equality of floating point Vectors, consider using `isclose`
+        """
         if type(other) is not self.__class__:
             return False
-        if self.dtype != other.dtype:
+        if check_dtype and self.dtype != other.dtype:
             return False
         if self.size != other.size:
             return False
         if self.nvals != other.nvals:
             return False
-        # Use ewise_mult to compare equality via intersection
+        if check_dtype:
+            # dtypes are equivalent, so not need to unify
+            common_dtype = self.dtype
+        else:
+            common_dtype = dtypes.unify(self.dtype, other.dtype)
+
         matches = Vector.new_from_type(bool, self.size)
-        matches << self.ewise_mult(other, binary.eq)
+        matches << self.ewise_mult(other, binary.eq[common_dtype])
+        # ewise_mult performs intersection, so nvals will indicate mismatched empty values
         if matches.nvals != self.nvals:
             return False
+
+        # Check if all results are True
+        result = Scalar.new_from_type(bool)
+        result << matches.reduce(monoid.land)
+        return result.value
+
+    def isclose(self, other, rtol=1e-7, atol=0.0, *, check_dtype=False):
+        """
+        Check for approximate equality (including same size and empty values)
+        If `check_dtype` is True, also checks that dtypes match
+        Closeness check is equivalent to `abs(a-b) <= max(rtol * max(abs(a), abs(b)), atol)`
+        """
+        if type(other) is not self.__class__:
+            return False
+        if check_dtype and self.dtype != other.dtype:
+            return False
+        if self.size != other.size:
+            return False
+        if self.nvals != other.nvals:
+            return False
+        if check_dtype:
+            # dtypes are equivalent, so not need to unify
+            common_dtype = self.dtype
+        else:
+            common_dtype = dtypes.unify(self.dtype, other.dtype)
+
+        matches = Vector.new_from_type(bool, self.size)
+        tmp1 = self.apply(unary.abs).new(dtype=common_dtype)
+        tmp2 = other.apply(unary.abs).new(dtype=common_dtype)
+        tmp1 << tmp1.ewise_mult(tmp2, monoid.max)
```

Comment:
Hmm, I bet these two lines (76-77) could be replaced with
```python
        tmp1(mask=other.S, accum=binary.max, replace=True) << other.apply(unary.abs)
```
Is this better?  I don't have much intuition yet on performance or behavior.  We would still need the second temporary to calculate `abs(a - b)`.

Reply:
Yeah, this would be more efficient. Let me update it.

Reply:
We could then define `tmp2` below via
```python
        tmp2 = self.ewise_mult(other, binary.minus).new(dtype=common_dtype, mask=tmp1.S)
```
Once again, I don't have much intuition here.  Providing the mask here is unnecessary, but will it help or hinder performance?

Reply:
Oh wait, this won't work. If an element only exists in `other`, it will exist in `tmp1`. To make this work, we would need a mask of `self.S & other.S`, which probably defeats the effort of avoiding tmp2.

Reply:
Oh, right.  Is there a test that would catch this?

Reply:
Just added tests for this